Image types enabled by additionalSupportedImageTypes do not show up in the Accept: header

litherum opened the pull request:
#### 151a757aa72b7bb9165f9c72fea85bbd09a9181d
<pre>
Image types enabled by additionalSupportedImageTypes do not show up in the Accept: header
<a href="https://bugs.webkit.org/show_bug.cgi?id=253046">https://bugs.webkit.org/show_bug.cgi?id=253046</a>
rdar://106006530

Reviewed by Said Abou-Hallawa.

We can simply augment the Accept: header with the mime types of these image formats.

* LayoutTests/http/tests/misc/heic-accept-header-expected.txt: Added.
* LayoutTests/http/tests/misc/heic-accept-header.html: Added.
* LayoutTests/http/tests/misc/resources/image-heic-accept.py: Added.
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::acceptHeaderValueForAdditionalSupportedImageMIMETypes):
(WebCore::acceptHeaderValueForImageResource):

Canonical link: <a href="https://commits.webkit.org/261131@main">https://commits.webkit.org/261131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8ac296b7e26e1fda5cd065f86b2cb4b9452762d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110688 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19769 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43251 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2026 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114656 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21178 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10881 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1548 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116429 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103004 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44114 "") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12953 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18326 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7708 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/14891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on LayoutTests/http/tests/misc/heic-accept-header.html:
```diff
@@ -0,0 +1,29 @@
+        if (image.width == 128 && image.height == 128)
+            result.innerHTML = "PASS";
+        else
+            result.innerHTML = "FAIL";
```
Should not we check whether the mime type of the image is `image/heic` instead of checking the image size? I am not sure how we check the mime type of an `HTMLImageElement` though? Maybe an internal API?

litherum replied:
Right, that's why it checks image size. The logic for this test is done in the server script.

shallawa commented on LayoutTests/http/tests/misc/resources/image-heic-accept.py:
```diff
@@ -0,0 +1,19 @@
+
+if 'image/heic' in http_accept:
+    sys.stdout.write(
+        'Content-Type: image/jpg\r\n'
```
I think we should simulate what the web server does and send a heic image.

shallawa commented on LayoutTests/http/tests/misc/resources/image-heic-accept.py:
```diff
@@ -0,0 +1,19 @@
+    with open(os.path.join(os.path.dirname(__file__), 'compass.jpg'), 'rb') as fn:
+        sys.stdout.buffer.write(fn.read())
+else:
+    sys.stdout.write('Content-Type: text/html\r\n\r\n')
```
Maybe we need to send the jpg image here since the accept header does not include heic.

shallawa commented on LayoutTests/http/tests/misc/heic-accept-header.html:
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE html><!-- webkit-test-runner [ additionalSupportedImageTypes=public.heic;public.heics ] -->
```
I think this test should be enabled only for macOS and iOS only since additionalSupportedImageTypes() is defined for Cocoa ports. And other ports do not support HEIC.

litherum replied:
🆗

litherum committed:
Image types enabled by additionalSupportedImageTypes do not show up in the Accept: header
https://bugs.webkit.org/show_bug.cgi?id=253046
rdar://106006530

Reviewed by Said Abou-Hallawa.

We can simply augment the Accept: header with the mime types of these image formats.

* LayoutTests/http/tests/misc/heic-accept-header-expected.txt: Added.
* LayoutTests/http/tests/misc/heic-accept-header.html: Added.
* LayoutTests/http/tests/misc/resources/image-heic-accept.py: Added.
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::acceptHeaderValueForAdditionalSupportedImageMIMETypes):
(WebCore::acceptHeaderValueForImageResource):

Canonical link: https://commits.webkit.org/261131@main
